prow/gerrit/adapter/adapter.go

```diff
@@ -91,54 +91,38 @@ func NewController(ctx context.Context, prowJobClient prowv1.ProwJobInterface, o
 	if err != nil {
 		logrus.WithError(err).Fatal("Error creating gerrit client.")
 	}
+	gerritClient.Authenticate(cookiefilePath, tokenPathOverride)
+
 	c := &Controller{
 		prowJobClient:      prowJobClient,
 		config:             cfg,
 		gc:                 gerritClient,
 		tracker:            lastSyncTracker,
 		projectsOptOutHelp: projectsOptOutHelpMap,
 	}
-
-	// applyGlobalConfig reads gerrit configurations from global gerrit config,
-	// it will completely override previously configured gerrit hosts and projects.
-	// it will also by the way authenticate gerrit
-	c.applyGlobalConfig(cfg, gerritClient, lastSyncTracker, cookiefilePath, tokenPathOverride)
-
-	return c
-}
-
-func (c *Controller) applyGlobalConfig(cfg config.Getter, gerritClient *client.Client, lastSyncTracker *syncTime, cookiefilePath, tokenPathOverride string) {
-	c.applyGlobalConfigOnce(cfg, gerritClient, lastSyncTracker, cookiefilePath, tokenPathOverride)
-
 	go func() {
 		for {
-			c.applyGlobalConfigOnce(cfg, gerritClient, lastSyncTracker, cookiefilePath, tokenPathOverride)
+			orgReposConfig := cfg().Gerrit.OrgReposConfig
+			if orgReposConfig == nil {
+				time.Sleep(time.Second)
+				continue
+			}
+			// Use globally defined gerrit repos if present
+			if err := gerritClient.UpdateClients(orgReposConfig.AllRepos()); err != nil {
+				logrus.WithError(err).Error("Updating clients.")
+			}
+			if err := lastSyncTracker.update(orgReposConfig.AllRepos()); err != nil {
+				logrus.WithError(err).Error("Syncing states.")
+			}
+			// Updates a map, lock to make sure it's thread safe.
+			c.lock.Lock()
+			c.projectsOptOutHelp = orgReposConfig.OptOutHelpRepos()
+			c.lock.Unlock()
 			// No need to spin constantly, give it a break. It's ok that config change has one second delay.
 			time.Sleep(time.Second)
 		}
 	}()
-}
-
-func (c *Controller) applyGlobalConfigOnce(cfg config.Getter, gerritClient *client.Client, lastSyncTracker *syncTime, cookiefilePath, tokenPathOverride string) {
-	orgReposConfig := cfg().Gerrit.OrgReposConfig
-	if orgReposConfig == nil {
-		return
```

Comment:
The root cause was that it returned here before calling the first Authenticate. Will fix in a following PR

Reply:
:man_facepalming: